### ISSUE TITLE
Proposal: Manual plugin triggers (decision logs, bundles, status, and discovery)

### ISSUE THREAD

**Issue by gshively11** [open-policy-agent/opa#2899]
Currently, OPA plugins run on a configureable timer, such that an opa instance will periodically attempt to download new bundles/discovery, and push status/decision logs. This works well for execution environments that have continual compute available (which is common).

On serverless compute platforms (e.g. AWS Lambda), it's not unusual to have execution environments not allow background processing and freeze when not in use. This makes the timer loops in the plugins very unreliable, because the timer/loop freezes when the execution environment freezes.  Consider this example:

1. An HTTP request triggers a lambda function
2. The execution environment unfreezes to handle the request
3. OPA plugin, running in a separate process in the execution environment, may be anywhere in its timer loop
4. The lambda handler finishes processing the http request and returns the response
5. The execution environment freezes
6. OPA plugin may still be in the timer loop, or may be in the middle of processing plugin actions. It has no way to keep the execution environment running while it finishes.

At the moment, there is a new AWS Lambda feature in preview: [Extensions](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-extensions-in-preview/). This would allow the opa server to control when the execution environment is released/frozen, at the expense of delaying the response to the client calling the lambda. When you consider the timer loops that opa users for the plugins, this isn't a feasible approach, because a timer may be configured at a long interval, or might be early in its timer delay.  

However, if there was as way to trigger the opa plugins to run on demand, we could have them triggered as soon as the execution environment unfreezes, and then have a way to listen for the response, which when received, would allow us to release/freeze the execution environment.  Ideally, we could still configure the plugins to only run on a manual trigger if X amount of time had passed (sort of like the timers, except instead of min/max interval, it would be an "atleast x amount of time since last trigger" interval).

Something like this could be accomplished in a couple of ways.  In our case, we have direct access to the Runtime (and thus the plugin Manager), because we are wrapping the opa server in our own go executable.  We could do something like `runtime.Manager.Plugin("bundle").Trigger()` (psuedo-code) and have that wait for the response, or have a listener or something.  The other idea, which @tsandall mentioned in slack, is to have routes on the OPA server which could be used to kick off plugins.  This is a more generic, integration-friendly approach I think, although it's less ideal for our specific case, since we don't want to run a separate process that needs to establish a tcp connection to trigger a plugin.  Maybe there is a way to handle both situations?  Or maybe there's another way to trigger plugins?

**Comment by patrick-east** [comment]
I kind of like the idea of having that `Trigger()` API for the plugins. We could make it a new optional interface built on top of the normal Plugin ones. That would give the golang users of OPA a nice solution.

We could, as needed, extend the server to support API's which turn around and use those too. Seems like it would be a pretty nice way to iterate on this after providing the golang API.

**Comment by zacyang** [comment]
Does this mean even in a container environment the plugin is not totally reliable as the container could be terminated at anytime as well like lambda just less frequently.

**Comment by gshively11** [comment]
@zacyang no, assuming you're referring to a typical container infrastructure (e.g. kubernetes).  In Lambda, the execution environment freezes, such that no processing happens at all until it's unfrozen, whereas something like a container running on kubernetes will always be processing until the container is terminated permanently.  

The fact that containers can be terminated doesn't make the loop-based plugins less reliable.  Typically, containers will send signals to their processes to indicate the container is shutting down, and the apps can flush any buffers and shut down gracefully, which things like the opa decision logger does.  If a container were to terminate ungracefully, there's nothing the opa plugins can do about that (except persist buffers to disk or something more permanent).

This issue is specific to lambda, or any other execution environments that don't have constant processing.

**Comment by tsandall** [comment]
@ashutosh-narkar is planning to pick this up in the coming days. This may land in v0.31.0 but may be pushed out to v0.32.0 to ensure stability.

**Comment by gshively11** [comment]
@tsandall Great news!  @ashutosh-narkar I'd love to help test out the changes before they go live. I already have a sample lambda project stood up that I used to test my original branch.

**Comment by ashutosh-narkar** [comment]
Hey @gshively11 can you share a link to your sample lambda project ? It would help during testing. Thanks !

**Comment by gshively11** [comment]
Unfortunately it's on our private GitHub and it's using a bunch of custom stuff that isn't opensource, so I can't.  I can try to pull together some snippets for you though, if that would be helpful.

**Comment by ashutosh-narkar** [comment]
That would help. I appreciate it. Thanks!
